File: mobile/app/src/main/java/com/ticketbus/mobile/offline/OfflineManager.java
package com.ticketbus.mobile.offline;

import android.content.Context;

import androidx.work.Constraints;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import com.ticketbus.mobile.data.local.AppDatabase;
import com.ticketbus.mobile.data.local.entity.CachedTicket;
import com.ticketbus.mobile.data.local.entity.PendingValidation;
import com.ticketbus.mobile.util.Constants;
import com.ticketbus.mobile.util.NetworkMonitor;
import com.ticketbus.mobile.util.PreferenceManager;

import java.util.concurrent.TimeUnit;

public class OfflineManager {
    private final Context context;
    private final AppDatabase database;
    private final PreferenceManager prefs;
    private final NetworkMonitor networkMonitor;

    public OfflineManager(Context context, AppDatabase database,
                          PreferenceManager prefs, NetworkMonitor networkMonitor) {
        this.context = context;
        this.database = database;
        this.prefs = prefs;
        this.networkMonitor = networkMonitor;
    }

    public boolean isOnline() {
        return networkMonitor.isConnected();
    }

    public int getPendingCount() {
        return database.validationEventDao().getUnsyncedCount();
    }

    public int getBlacklistCount() {
        return database.blacklistDao().count();
    }

    public void schedulePeriodicSync() {
        Constraints constraints = new Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build();
        PeriodicWorkRequest req = new PeriodicWorkRequest.Builder(
            SyncWorker.class, Constants.SYNC_INTERVAL_MINUTES, TimeUnit.MINUTES)
            .setConstraints(constraints)
            .build();
        WorkManager.getInstance(context)
            .enqueueUniquePeriodicWork(
                Constants.SYNC_WORK_NAME,
                ExistingPeriodicWorkPolicy.KEEP,
                req
            );
    }

    public void triggerImmediateSync() {
        Constraints constraints = new Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build();
        OneTimeWorkRequest req = new OneTimeWorkRequest.Builder(SyncWorker.class)
            .setConstraints(constraints)
            .build();
        WorkManager.getInstance(context).enqueue(req);
    }

    public boolean isBlacklisted(String ticketNumber) {
        return database.blacklistDao().findByTicketNumber(ticketNumber) != null;
    }

    public boolean hasSeenNonce(String ticketNumber) {
        return database.ticketCacheDao().findByTicketNumber(ticketNumber) != null;
    }

    public void recordValidation(String ticketNumber, String nonce,
                                  String deviceId, String location, String result) {
        database.ticketCacheDao().insert(new CachedTicket(ticketNumber, nonce));
        database.validationEventDao().insert(
            new PendingValidation("", deviceId, location, null, null, result)
        );
    }
}

File: mobile/app/src/main/java/com/ticketbus/mobile/offline/OfflineManager.kt
package com.ticketbus.mobile.offline

import android.content.Context
import androidx.work.*
import com.ticketbus.mobile.data.local.AppDatabase
import com.ticketbus.mobile.data.local.entity.PendingValidation
import com.ticketbus.mobile.util.Constants
import com.ticketbus.mobile.util.NetworkMonitor
import com.ticketbus.mobile.util.PreferenceManager
import java.util.concurrent.TimeUnit

class OfflineManager(
    private val context: Context,
    private val database: AppDatabase,
    private val prefs: PreferenceManager,
    private val networkMonitor: NetworkMonitor
) {
    fun isOnline() = networkMonitor.isConnected()
    suspend fun getPendingCount() = database.validationEventDao().getUnsyncedCount()
    suspend fun getBlacklistCount() = database.blacklistDao().count()

    fun schedulePeriodicSync() {
        val req = PeriodicWorkRequestBuilder<SyncWorker>(Constants.SYNC_INTERVAL_MINUTES, TimeUnit.MINUTES)
            .setConstraints(Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build())
            .build()
        WorkManager.getInstance(context)
            .enqueueUniquePeriodicWork(Constants.SYNC_WORK_NAME, ExistingPeriodicWorkPolicy.KEEP, req)
    }

    fun triggerImmediateSync() {
        val req = OneTimeWorkRequestBuilder<SyncWorker>()
            .setConstraints(Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build())
            .build()
        WorkManager.getInstance(context).enqueue(req)
    }

    suspend fun isBlacklisted(ticketNumber: String) =
        database.blacklistDao().findByTicketNumber(ticketNumber) != null

    suspend fun hasSeenNonce(ticketNumber: String) =
        database.ticketCacheDao().findByTicketNumber(ticketNumber) != null

    suspend fun recordValidation(ticketNumber: String, nonce: String, deviceId: String, location: String, result: String) {
        database.ticketCacheDao().insert(com.ticketbus.mobile.data.local.entity.CachedTicket(ticketNumber, nonce))
        database.validationEventDao().insert(
            PendingValidation(qrData = "", deviceId = deviceId, location = location, latitude = null, longitude = null, result = result)
        )
    }
}
